R/nimble_lm.R

```diff
@@ -5,40 +5,152 @@ library(assertthat) ## to package dependencies
+## These Functions Make the Name someFactor.X
 make.effect.name <- function(term) {
-  paste0(term, '.effect')
+  as.name(paste0(term, '.effect'))
 }
 
+make.mean.name <- function(term) {
+  as.name(paste0(term, '.mu'))
+}
+
+make.precision.name <- function(term) {
+  as.name(paste0(term, '.tau'))
+}
+
+make.sigma.name <- function(term) {
+  as.name(paste0(term, '.sigma2'))
+}
+
+
+
 #Make Factor Variable
 makeFac <- function(x,char.only=FALSE) {
   if (!is.factor(x) && (!char.only || is.character(x))) factor(x) else x
 }
 
 
+#Extracts formula From Environment
+RHSForm <- function(form,as.form=FALSE) {
+  rhsf <- form[[length(form)]]
+  if (as.form) reformulate(deparse(rhsf)) else rhsf
+}
+
+
+#Functions for Random Effects
+
+#Replaces '|' with '+'
+
+subbars <- function(term)
+{
+  if (is.name(term) || !is.language(term)) return(term)
+  if (length(term) == 2) {
+    term[[2]] <- subbars(term[[2]])
+    return(term)
+  }
+  stopifnot(length(term) >= 3)
+  if (is.call(term) && term[[1]] == as.name('|'))
+    term[[1]] <- as.name('+')
+  for (j in 2:length(term)) term[[j]] <- subbars(term[[j]])
+  term
+}
+
+#Find Random Effect Terms from Formula object
+fb <- function(term)
+{
+  if (is.name(term) || !is.language(term)) return(NULL)
+  if (term[[1]] == as.name("(")) return(fb(term[[2]]))
+  stopifnot(is.call(term))
+  if (term[[1]] == as.name('|')) return(term)
+  if (length(term) == 2) return(fb(term[[2]]))
+  c(fb(term[[2]]), fb(term[[3]]))
+}
+
+#Find Fixed Effect Terms
+nobars <- function(term) {
+  nb <- nobars_(term)  ## call recursive version
+  ## called with one-sided RE-only formula, or RHS alone
+  if (is.null(nb)) {
+    nb <- if (is(term,"formula")) ~1 else 1
+  }
+  nb
+}
+
+
+nobars_ <- function(term)
+{
+  if (!anyBars(term)) return(term)
+  if (isBar(term)) return(NULL)
+  if (isAnyArgBar(term)) return(NULL)
+  if (length(term) == 2) {
+    nb <- nobars_(term[[2]])
+    if(is.null(nb)) return(NULL)
+    term[[2]] <- nb
+    return(term)
+  }
+  nb2 <- nobars_(term[[2]])
+  nb3 <- nobars_(term[[3]])
+  if (is.null(nb2)) return(nb3)
+  if (is.null(nb3)) return(nb2)
+  term[[2]] <- nb2
+  term[[3]] <- nb3
+  term
+}
+
+isBar <- function(term) {
+  if(is.call(term)) {
+    if((term[[1]] == as.name("|"))) {
+      return(TRUE)
+    }
+  }
+  FALSE
+}
+
+isAnyArgBar <- function(term) {
+  if ((term[[1]] != as.name("~")) && (term[[1]] != as.name("("))) {
+    for(i in seq_along(term)) {
+      if(isBar(term[[i]])) return(TRUE)
+    }
+  }
+  FALSE
+}
+
+anyBars <- function(term) {
+  any(c('|','||') %in% all.names(term))
+}
+
 
 
 #Make Normal Diffuse Prior N ~ (0, 100,000), Uniform (0,1000), t(0,1,d), or point mass at 0
 #We Could Also Allow an Argument to Specify the Shape and Scale Parameter?
-make.prior <- function(term,prior,...){
+make.prior <- function(term, prior,...){
 
   if (prior=="Normal"){
-  substitute(TERM ~ dnorm(0,0.0001), list(TERM=term))
+  substitute(TERM ~ dnorm(0, 0.0001), list(TERM=term))
```

Comment:
Since NIMBLE supports alternative parameterizations, we've preferred to use `dnorm(0, sd = 1000)`, i.e. sd instead of precision.  This moots the need for a separate declaration to get sd from precision.